File: lib/services/affiliate_service.dart
import 'package:supabase_flutter/supabase_flutter.dart';

import '../core/supabase_client.dart';

/// Port of affiliate/services/affiliate.service.ts.

class Referral {
  const Referral({
    required this.id,
    required this.name,
    required this.active,
    required this.joinedAt,
  });

  final String id;
  final String name;
  final bool active;
  final DateTime? joinedAt;
}

class Earning {
  const Earning({
    required this.id,
    required this.name,
    required this.plan,
    required this.amount,
    required this.commission,
    required this.createdAt,
  });

  final String id;
  final String name;
  final String plan;
  final double amount;
  final double commission;
  final DateTime? createdAt;

  String get planLabel => plan.isEmpty
      ? 'Subscription'
      : '${plan[0].toUpperCase()}${plan.substring(1)} Plan';
}

enum WithdrawalStatus { completed, processing, failed }

class Withdrawal {
  const Withdrawal({
    required this.id,
    required this.amount,
    required this.status,
    required this.phone,
    required this.createdAt,
  });

  final String id;
  final double amount;
  final WithdrawalStatus status;
  final String phone;
  final DateTime? createdAt;
}

class AffiliateSummary {
  const AffiliateSummary({
    required this.enrolled,
    required this.code,
    required this.since,
    required this.availableBalance,
    required this.pendingBalance,
    required this.totalEarned,
    required this.referrals,
    required this.earnings,
    required this.withdrawals,
    required this.earningsToday,
    required this.earningsThisWeek,
    required this.earningsThisMonth,
  });

  const AffiliateSummary.empty()
    : enrolled = false,
      code = '',
      since = null,
      availableBalance = 0,
      pendingBalance = 0,
      totalEarned = 0,
      referrals = const [],
      earnings = const [],
      withdrawals = const [],
      earningsToday = 0,
      earningsThisWeek = 0,
      earningsThisMonth = 0;

  final bool enrolled;
  final String code;
  final DateTime? since;
  final double availableBalance;
  final double pendingBalance;
  final double totalEarned;
  final List<Referral> referrals;
  final List<Earning> earnings;
  final List<Withdrawal> withdrawals;
  final double earningsToday;
  final double earningsThisWeek;
  final double earningsThisMonth;

  int get referralsCount => referrals.length;
  int get activeUsersCount => referrals.where((r) => r.active).length;
  int get conversionRate => referrals.isEmpty
      ? 0
      : ((activeUsersCount / referrals.length) * 100).round();
}

class AffiliateService {
  const AffiliateService();

  static const minWithdrawal = 100.0;

  Future<AffiliateSummary> load() async {
    final uid = currentUserId;
    if (uid == null) return const AffiliateSummary.empty();

    final affiliate = await supabase
        .from('affiliates')
        .select()
        .eq('id', uid)
        .maybeSingle();

    if (affiliate == null) return const AffiliateSummary.empty();

    // RLS scopes each of these to this affiliate (users_select allows reading
    // the users you referred; commissions_select and payout_requests_select
    // match on affiliate_id), so the filters are for clarity, not security.
    final results = await Future.wait([
      supabase
          .from('users')
          .select('id, name, is_subscribed, created_at')
          .eq('referred_by_uid', uid),
      supabase
          .from('commissions')
          .select('id, referred_id, plan, amount, commission_amount, created_at')
          .eq('affiliate_id', uid)
          .order('created_at', ascending: false),
      supabase
          .from('payout_requests')
          .select('id, amount, phone, status, created_at')
          .eq('affiliate_id', uid)
          .order('created_at', ascending: false),
    ]);

    final nameByUid = <String, String>{};
    final referrals = results[0].map((row) {
      final name = row['name'] as String? ?? 'AfyaSmart User';
      nameByUid[row['id'] as String] = name;
      return Referral(
        id: row['id'] as String,
        name: name,
        active: row['is_subscribed'] as bool? ?? false,
        joinedAt: _parse(row['created_at']),
      );
    }).toList();

    final now = DateTime.now();
    final startOfToday = DateTime(now.year, now.month, now.day);
    final startOfWeek = startOfToday.subtract(
      Duration(days: now.weekday == DateTime.sunday ? 6 : now.weekday - 1),
    );
    final startOfMonth = DateTime(now.year, now.month);

    var today = 0.0;
    var week = 0.0;
    var month = 0.0;

    final earnings = results[1].map((row) {
      final commission = _toDouble(row['commission_amount']);
      final createdAt = _parse(row['created_at']);

      if (createdAt != null) {
        if (!createdAt.isBefore(startOfToday)) today += commission;
        if (!createdAt.isBefore(startOfWeek)) week += commission;
        if (!createdAt.isBefore(startOfMonth)) month += commission;
      }

      return Earning(
        id: row['id'] as String,
        name: nameByUid[row['referred_id']] ?? 'Referred user',
        plan: row['plan'] as String? ?? '',
        amount: _toDouble(row['amount']),
        commission: commission,
        createdAt: createdAt,
      );
    }).toList();

    final withdrawals = results[2]
        .map(
          (row) => Withdrawal(
            id: row['id'] as String,
            amount: _toDouble(row['amount']),
            status: switch (row['status']) {
              'paid' => WithdrawalStatus.completed,
              'rejected' => WithdrawalStatus.failed,
              _ => WithdrawalStatus.processing,
            },
            phone: row['phone'] as String? ?? '',
            createdAt: _parse(row['created_at']),
          ),
        )
        .toList();

    return AffiliateSummary(
      enrolled: true,
      code: affiliate['code'] as String? ?? '',
      since: _parse(affiliate['created_at']),
      availableBalance: _toDouble(affiliate['available_balance']),
      pendingBalance: _toDouble(affiliate['pending_balance']),
      totalEarned: _toDouble(affiliate['total_earned']),
      referrals: referrals,
      earnings: earnings,
      withdrawals: withdrawals,
      earningsToday: today,
      earningsThisWeek: week,
      earningsThisMonth: month,
    );
  }

  /// Claims a referral code. The RPC settles collisions against the unique
  /// index and returns the existing code if already enrolled.
  Future<String> enroll() async {
    try {
      final code = await supabase.rpc('enroll_affiliate');
      return code as String? ?? '';
    } on PostgrestException catch (e) {
      throw ApiException(
        e.message.isEmpty
            ? 'Could not enroll as an affiliate. Please try again.'
            : e.message,
      );
    }
  }

  /// Requests a payout.
  ///
  /// The guards here are for a fast, friendly message. request_payout
  /// re-checks both authoritatively and deducts the balance under a row lock,
  /// so two concurrent withdrawals cannot both succeed.
  Future<void> requestPayout({
    required double amount,
    required String phone,
    required double availableBalance,
  }) async {
    if (amount <= 0 || amount < minWithdrawal) {
      throw const ApiException(
        'Minimum withdrawal amount is Ksh 100.',
      );
    }
    if (amount > availableBalance) {
      throw const ApiException('You do not have enough available balance.');
    }

    try {
      await supabase.rpc(
        'request_payout',
        params: {'p_amount': amount, 'p_phone': phone},
      );
    } on PostgrestException catch (e) {
      throw ApiException(
        e.message.isEmpty ? 'Unable to process withdrawal request.' : e.message,
      );
    }
  }

  static DateTime? _parse(Object? value) =>
      value is String ? DateTime.tryParse(value)?.toLocal() : null;

  static double _toDouble(Object? value) => switch (value) {
    final num n => n.toDouble(),
    final String s => double.tryParse(s) ?? 0,
    _ => 0,
  };
}

File: affiliate/services/affiliate.service.ts
import { create } from 'zustand';
import { supabase, getCurrentUserId } from '@/src/services/supabase';

export interface Referral {
  id: string;
  name: string;
  status: 'Active' | 'Inactive';
  joinedDate: string;
}

export interface Earning {
  id: string;
  name: string;
  planName: string;
  amount: number;
  commission: number;
  date: string;
}

export interface Withdrawal {
  id: string;
  amount: number;
  date: string;
  status: 'Completed' | 'Processing' | 'Failed';
  mpesaNumber: string;
}

const formatDate = (value: string | null | undefined): string => {
  const date = value ? new Date(value) : null;
  if (!date || Number.isNaN(date.getTime())) return 'Recent';
  return new Intl.DateTimeFormat('en-KE', {
    day: '2-digit',
    month: 'short',
    year: 'numeric',
    hour: '2-digit',
    minute: '2-digit',
  }).format(date);
};

const planLabel = (plan?: string) =>
  plan ? `${plan.charAt(0).toUpperCase()}${plan.slice(1)} Plan` : 'Subscription';

const payoutStatusLabel = (status: string): Withdrawal['status'] => {
  if (status === 'paid') return 'Completed';
  if (status === 'rejected') return 'Failed';
  return 'Processing';
};

interface AffiliateState {
  loading: boolean;
  enrolled: boolean;
  affiliateId: string;
  affiliateSince: string | null;
  availableBalance: number;
  pendingEarnings: number;
  totalEarned: number;
  referralsCount: number;
  activeUsersCount: number;
  conversionRate: number;
  referrals: Referral[];
  earnings: Earning[];
  withdrawals: Withdrawal[];
  earningsToday: number;
  earningsThisWeek: number;
  earningsThisMonth: number;
  load: () => Promise<void>;
  enroll: () => Promise<{ success: boolean; message?: string }>;
  addWithdrawal: (
    amount: number,
    mpesaNumber: string,
  ) => Promise<{ success: boolean; message?: string }>;
}

const emptyState = {
  loading: false,
  enrolled: false,
  affiliateId: '',
  affiliateSince: null as string | null,
  availableBalance: 0,
  pendingEarnings: 0,
  totalEarned: 0,
  referralsCount: 0,
  activeUsersCount: 0,
  conversionRate: 0,
  referrals: [] as Referral[],
  earnings: [] as Earning[],
  withdrawals: [] as Withdrawal[],
  earningsToday: 0,
  earningsThisWeek: 0,
  earningsThisMonth: 0,
};

export const useAffiliateStore = create<AffiliateState>((set, get) => ({
  ...emptyState,

  load: async () => {
    const uid = await getCurrentUserId();
    if (!uid) {
      set({ ...emptyState });
      return;
    }

    set({ loading: true });

    const { data: affiliate } = await supabase
      .from('affiliates')
      .select('*')
      .eq('id', uid)
      .maybeSingle();

    if (!affiliate) {
      set({ ...emptyState, loading: false });
      return;
    }

    // RLS scopes each of these to this affiliate (users_select allows reading
    // the users you referred; commissions_select and payout_requests_select
    // match on affiliate_id), so the filters are for clarity, not security.
    const [referralsRes, earningsRes, payoutsRes] = await Promise.all([
      supabase
        .from('users')
        .select('id, name, is_subscribed, created_at')
        .eq('referred_by_uid', uid),
      supabase
        .from('commissions')
        .select('id, referred_id, plan, amount, commission_amount, created_at')
        .eq('affiliate_id', uid)
        .order('created_at', { ascending: false }),
      supabase
        .from('payout_requests')
        .select('id, amount, phone, status, created_at')
        .eq('affiliate_id', uid)
        .order('created_at', { ascending: false }),
    ]);

    const referredNameByUid = new Map<string, string>();
    const referrals: Referral[] = (referralsRes.data ?? []).map((row: any) => {
      const name = row.name ?? 'AfyaSmart User';
      referredNameByUid.set(row.id, name);
      return {
        id: row.id,
        name,
        status: row.is_subscribed ? 'Active' : 'Inactive',
        joinedDate: formatDate(row.created_at),
      };
    });

    const activeUsersCount = referrals.filter((r) => r.status === 'Active').length;
    const conversionRate =
      referrals.length > 0 ? Math.round((activeUsersCount / referrals.length) * 100) : 0;

    const now = new Date();
    const startOfToday = new Date(now.getFullYear(), now.getMonth(), now.getDate());
    const dayOfWeek = now.getDay();
    const startOfWeek = new Date(startOfToday);
    startOfWeek.setDate(startOfToday.getDate() - (dayOfWeek === 0 ? 6 : dayOfWeek - 1));
    const startOfMonth = new Date(now.getFullYear(), now.getMonth(), 1);

    let earningsToday = 0;
    let earningsThisWeek = 0;
    let earningsThisMonth = 0;

    const earnings: Earning[] = (earningsRes.data ?? []).map((row: any) => {
      const commission = Number(row.commission_amount ?? 0);
      const createdAt = row.created_at ? new Date(row.created_at) : null;

      if (createdAt && !Number.isNaN(createdAt.getTime())) {
        if (createdAt >= startOfToday) earningsToday += commission;
        if (createdAt >= startOfWeek) earningsThisWeek += commission;
        if (createdAt >= startOfMonth) earningsThisMonth += commission;
      }

      return {
        id: row.id,
        name: referredNameByUid.get(row.referred_id) ?? 'Referred user',
        planName: planLabel(row.plan),
        amount: Number(row.amount ?? 0),
        commission,
        date: formatDate(row.created_at),
      };
    });

    const withdrawals: Withdrawal[] = (payoutsRes.data ?? []).map((row: any) => ({
      id: row.id,
      amount: Number(row.amount ?? 0),
      date: formatDate(row.created_at),
      status: payoutStatusLabel(row.status ?? 'pending'),
      mpesaNumber: row.phone ?? '',
    }));

    set({
      loading: false,
      enrolled: true,
      affiliateId: affiliate.code ?? '',
      affiliateSince: formatDate(affiliate.created_at),
      availableBalance: Number(affiliate.available_balance ?? 0),
      pendingEarnings: Number(affiliate.pending_balance ?? 0),
      totalEarned: Number(affiliate.total_earned ?? 0),
      referralsCount: referrals.length,
      activeUsersCount,
      conversionRate,
      referrals,
      earnings,
      withdrawals,
      earningsToday,
      earningsThisWeek,
      earningsThisMonth,
    });
  },

  enroll: async () => {
    const { error } = await supabase.rpc('enroll_affiliate');

    if (error) {
      return {
        success: false,
        message: error.message || 'Could not enroll as an affiliate. Please try again.',
      };
    }

    await get().load();
    return { success: true };
  },

  addWithdrawal: async (amount: number, mpesaNumber: string) => {
    const { availableBalance } = get();

    // Client-side guards for a fast, friendly message. request_payout
    // re-checks both authoritatively, and deducts the balance under a row
    // lock so two concurrent withdrawals cannot both succeed.
    if (amount <= 0 || amount < 100) {
      return { success: false, message: 'Minimum withdrawal amount is Ksh 100.' };
    }
    if (amount > availableBalance) {
      return { success: false, message: 'You do not have enough available balance.' };
    }

    const { error } = await supabase.rpc('request_payout', {
      p_amount: amount,
      p_phone: mpesaNumber,
    });

    if (error) {
      return {
        success: false,
        message: error.message || 'Unable to process withdrawal request.',
      };
    }

    await get().load();
    return { success: true };
  },
}));
